### DISCUSSION TITLE
Is there a way to find a workspace?

### DISCUSSION THREAD

**Discussion by landtuna** [jj-vcs/jj#5084] [Q&A] [answered]
It seems that `jj workspace list` doesn't tell you where a workspace is located. Is there a way to find out this information? I resorted to `find ~ -name repo -type f`.

**Comment by martinvonz** [comment, marked as answer]
No, there is no way. It's that way by design because the workspaces may live somewhere that the repo doesn't even know about, like on a USB stick. At Google, where repos are stored in a central database, there can still be workspaces on a user's local machine.

**Reply by landtuna** [reply]
Thanks!

**Reply by arxanas** [reply]
Is that a good OSS design?

- Realistically I'd probably lose my workspaces just like in the original post.
- It sounds like it'll make it hard for automation to programmatically manage workspaces.

Maybe an incrementally-better design is to mark workspaces as locally-accessible or not?

- Git also has the same issue where a worktree may reside on a networked filesystem, etc. so you have to `lock` it so that `prune` doesn't think it's missing and try to clean it up.
  - That is, it _knows about_ and _accommodates_ non-local workspaces in the design, but doesn't recognize them as a top-level concept.
- It would be nice to potentially mark workspaces as "programmatically-managed" as well, although that's a different issue.
- It's true that the management of bidirectional awareness of Git worktrees is annoying, unlike jj which I assume is unidirectional for now.
  - In Git, if you move a worktree without running the appropriate command, it'll get confused and you'll have to remediate.

I guess I could see unidirectionally-linked workspaces as a more tractable design, though.
- It might not be so bad to have users and tooling establish their own conventions for workspace location, if the alternative is the mess of administrative overhead of `git worktree`.

**Reply by martinvonz** [reply]
Is it much more likely that people forgetting where their additional workspaces are than forgetting where their main workspace is? Is it that people create temporary workspaces in `/tmp/` or something?

I don't mind adding some non-authoritative information about where workspaces may possibly be found.

**Reply by landtuna** [reply]
Yes, this is how I ended up in the situation. I make temporary workspaces sometimes when I'm doing two things at once. And then quite a while later I'll run `jj workspace list` and wonder what those random workspaces were. Every time, there's nothing important there, but I still want to check if maybe I left anything in the temporary workspace before I run `jj workspace forget`. (Though I guess only non-tracked files could be lying around in them.)

**Reply by martinvonz** [reply]
FYI, `jj workspace forget` doesn't actually delete any files, so maybe part of the answer is that we should make it possible to add back an existing workspace (in case you find a workspace later). I think we have a TODO somewhere about that.

**Reply by konradzapalowicz** [reply]
> Is it much more likely that people forgetting where their additional workspaces are than forgetting where their main workspace is? Is it that people create temporary workspaces in `/tmp/` or something?
> 
> I don't mind adding some non-authoritative information about where workspaces may possibly be found.

I literally forget where my workspaces live or more precisely which folder is which. Especially, these days, when I have a high number of them, with concurrent work streams, I struggle to find out which ones (folders) I can `rm` or recycle and which I cannot. 

Right now I have like 5 workspaces besides default and 8 folders named $project-1,2,3....,N Part of my issue is that the folders are not named after a workspace, which makes discovery harder. However, still, it would be great to be able to know which one is which.

**Reply by anka-213** [reply]
I have also forgotten where workspaces are several times and had to search my whole home-dir to find them. I would also appreciate some non-authorative information about possible workspace location.

**Reply by josephlou5** [reply]
Btw these days (jj v0.44.0), `jj workspace list` includes workspace roots by default when it's known (tracked by jj if the workspace is created after v0.38.0 I believe) (#9713, #9826).
